File: iteration2/Models/Transcript.java
package iteration2.Models;


import org.json.JSONArray;
import org.json.JSONObject;
import java.io.IOException;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.logging.FileHandler;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public class Transcript extends Model{
    private double gpa;
    private int completedCredit;
    private ArrayList<Course> completedCourses;
    private  ArrayList<Course> failedCourses;

    private static final String FILE_PATH="/iteration2/Logs/Transcript.log";
    private static Logger logger=Logger.getLogger(Transcript.class.getName());
    private static FileHandler fileHandler;

    static {
        try {
            fileHandler = new FileHandler(System.getProperty("user.dir")+FILE_PATH,true);
            logger.addHandler(fileHandler);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void customLog(boolean type,String message){
        SimpleFormatter formatter =new SimpleFormatter();
        fileHandler.setFormatter(formatter);
        if(type)
            logger.info(message);
        else
            logger.warning(message);
    }

    public Transcript(double gpa, int completedCredit, ArrayList<Course> completedCourses, ArrayList<Course> failedCourses) {
        this.gpa = gpa;
        this.completedCredit = completedCredit;
        this.completedCourses = completedCourses;
        this.failedCourses = failedCourses;
        customLog(true,"Transcript constructor method is called and new transcript object is generated.");
    }

    public void  calculateGPA(){
        double tempGpa=0;
        int totalCredit=0;
        int courseCredit=0;
        for (Course course:completedCourses){
            courseCredit=course.getCredit();
            totalCredit+=courseCredit;
            tempGpa+=courseCredit*4;
        }
        for (Course course:failedCourses){
            courseCredit=course.getCredit();
            totalCredit+=courseCredit;
            tempGpa+=courseCredit*0;
        }
        this.gpa=tempGpa/totalCredit;
        customLog(true,"The student's GPA is calculated.");
    }

    public double getGpa() {
        return gpa;
    }

    public void setGpa(double gpa) {
        if (gpa >= 0) {
            this.gpa = gpa;
        }
    }

    public int getCompletedCredit() {
        return completedCredit;
    }

    public void setCompletedCredit(int completedCredit) {
        if (completedCredit>=0){
            this.completedCredit = completedCredit;
        }
    }

    public ArrayList<Course> getCompletedCourses() {
        return completedCourses;
    }

    public void setCompletedCourses(ArrayList<Course> completedCourses) {
        for (int i=0; i<completedCourses.size();i++){
            for (int j = 0; j<this.completedCourses.size();j++){
                if (completedCourses.get(i).getCode()!=this.completedCourses.get(j).getCode()){
                    this.completedCourses.add(completedCourses.get(i));
                }
            }
        }
    }

    public ArrayList<Course> getFailedCourses() {
        return failedCourses;
    }

    public void setFailedCourses(ArrayList<Course> failedCourses) {
        for (int i=0; i<failedCourses.size();i++){
            for (int j = 0; j<this.failedCourses.size();j++){
                if (failedCourses.get(i).getCode()!=this.failedCourses.get(j).getCode()){
                    this.failedCourses.add(failedCourses.get(i));
                }
            }
        }
    }

    public void addToCompletedCourses(Course course){
        if (!completedCourses.contains(course)){
            completedCourses.add(course);
            customLog(true,"A new course has been added to the courses that the student has completed.");
        }else {
            customLog(false,"A new course could not be added to the courses that the student completed" +
                    " because it has already completed it.");
        }
    }

    public void addToFailedCourses(Course course){
        if (!failedCourses.contains(course)){
            failedCourses.add(course);
            customLog(true,"A new course has been added to the courses that the student has failed.");
        }else {
            customLog(false,"A new course could not be added to the courses that the student failed because it has already failed it.");
        }
    }

    public JSONArray convertArraylistToJsonArray(ArrayList<Course> arrayList){
        JSONArray jsonArray=new JSONArray();
        for (int i=0;i<arrayList.size();i++){
            jsonArray.put(arrayList.get(i).toJson());
        }
        return jsonArray;
    }

    @Override
    public JSONObject toJson() {
        JSONObject jsonObject=new JSONObject();
        jsonObject.put("gpa",new DecimalFormat("##.##").format(gpa));
        jsonObject.put("completedCredit",completedCredit);
        jsonObject.put("completedCourses",convertArraylistToJsonArray(completedCourses));
        jsonObject.put("failedCourses",convertArraylistToJsonArray(failedCourses));
        return  jsonObject;
    }
}

File: iteration3/Models/Transcript.py
import json
from decimal import Decimal

from iteration3.Models.Model import Model


class Transcript(Model):

    def __init__(self, gpa:float, completed_credit:int, completed_courses:[], failed_courses:[]):
        self.gpa = gpa
        self.completed_credit = completed_credit
        self.completed_courses = completed_courses
        self.failed_courses = failed_courses

    def calculate_gpa(self):
        letter_grade_dict = {
            "AA": 4.0,
            "BA": 3.5,
            "BB": 3.0,
            "BC": 2.5,
            "CC": 2.0,
            "DC": 1.5,
            "DD": 1.0,
            "FF": 0.0
        }
        gpa = 0
        credit = 0
        for i in range(len(self.completed_courses)):
            gpa += self.completed_courses[i].get_credit() * letter_grade_dict[self.completed_courses[i].get_letter_grade()]
            credit += self.completed_courses[i].get_credit()
        for i in range(len(self.failed_courses)):
            gpa += self.failed_courses[i].get_credit() * letter_grade_dict[self.failed_courses[i].get_letter_grade()]
            credit += self.failed_courses[i].get_credit()
        self.gpa = gpa / credit


    def get_gpa(self):
        return self.gpa

    def set_gpa(self, gpa:float):
        if gpa >= 0:
            self.gpa = gpa

    def get_completed_credit(self):
        return self.completed_credit

    def set_completed_credit(self, completed_credit:int):
        if completed_credit >= 0:
            self.completed_credit = completed_credit

    def get_completed_courses(self):
        return self.completed_courses

    def set_completed_courses(self, completed_courses:[]):
        for i in range(len(completed_courses)):
            for j in range(len(self.completed_courses)):
                if completed_courses[i].get_code() != self.completed_courses[j].get_code():
                    self.completed_courses.add(completed_courses[i])

    def get_failed_courses(self):
        return self.failed_courses

    def set_failed_courses(self, failed_courses:[]):
        for i in range(len(failed_courses)):
            for j in range(len(self.failed_courses)):
                if failed_courses[i].get_code() != self.failed_courses[j].get_code():
                    self.failed_courses.add(failed_courses[i])

    def add_to_completed_courses(self, course):
        if course not in self.completed_courses:
            self.completed_courses.add(course)

    def add_to_failed_courses(self, course):
        if course not in self.failed_courses:
            self.failed_courses.add(course)

    def convert_arraylist_to_json_array(self, array_list:[]):
        json_array = []
        for i in range(len(array_list)):
            json_array.append(array_list[i].to_json())
        return json_array

    def to_json(self):
        json_object = {}
        json_object['gpa'] = format(self.gpa, '##.##')
        json_object['completedCredit'] = self.completed_credit
        json_object['completedCourses'] = self.convert_arraylist_to_json_array(self.completed_courses)
        json_object['failedCourses'] = self.convert_arraylist_to_json_array(self.failed_courses)
        return json_object
